### DISCUSSION TITLE
setUpViewerAsEmbeddedInWindow help

### DISCUSSION THREAD

**Discussion by Mattc2021** [openscenegraph/OpenSceneGraph#1228] [Q&A]
Just had a quick question about the above function. If we have two GLFW contexts will setUpViewerAsEmbeddedInWindow embed to the current GLFW context we have or is it something different that I'm missing. In addition to this I was wondering if there was anything special I had to do for osgViewer->frame() with multiple GLFW contexts, it seems to come back with a RenderBin::Draw() openGL error every time.

**Comment by robertosfield** [comment]
setUpViewerAsEmbeddedInWindow() is just for single context single threaded applications, it does a non op for key parts of the GraphicsWindow that are required for multi-context work.

GLWF won't add anything over what the core osgViewer native windowing will provide so why not just uses the OSG out of the box, it save you lots of hassle.

If you really must use GLFW then have a look at the various subclasses of GraphicsWindow to see what they implement and create a GraphicsWindowFLFW equivalent.

**Reply by Mattc2021** [reply]
I'm adding an OpenSceneGraph (specifically OSGEarth) to an already existing project that utilizes GLFW windows. This new glfw window is being spawned off from a click of a button on an ImGui menu and to get it to work properly with being able to utilize both the already existing GLFW windows and the new OSGEarth window, then I had to find out a way to embed it in the new GLFW window. Thank you for the help though and I will look at the various subclasses of GraphicsWindow to see what I can do.

**Reply by Mattc2021** [reply]
More clarification:

We are only using osg for OSGEarth, there are no alternative C++ map viewers available AFAIK.  Our project already uses GLFW directly and low level OpenGL code with functionality that OSG does not provide.  We therefor cannot use OSG for the rest of our project, and we need to use OSG earth.  So the hope was that we could isolate the OSGEarth part of the code in it's own window, so it wouldn't interfere with the rest of our code base.

**Reply by robertosfield** [reply]
It should be possible but may not be entirely straight forward. osgEarth is written to be used as part of an OpenSceneGraph application, rather than a general purpose library that can be used in any OpenGL application with minimal work, it's not quite what you want it to be, so you'll need to do work to adapt your application to work with how the OSG/osgEarth work.  I have provided as many pointers as I can, the rest is to you.
